File: frontend/src/lib/bluesky.ts
import { useQuery, useMutation, useQueryClient } from '@tanstack/react-query';
import { api } from './api';
import { CanonicalPost, PlatformMapping } from './connectors';

// AT Protocol Types

export interface ATProtoDID {
  did: string;
  handle: string;
  pds: string;
  resolved_at: string;
}

export interface ATProtoSession {
  did: string;
  handle: string;
  accessJwt: string;
  refreshJwt: string;
  pds_url: string;
  created_at: string;
  expires_at: string;
}

export interface ATProtoProfile {
  did: string;
  handle: string;
  displayName?: string;
  description?: string;
  avatar?: string;
  banner?: string;
  followersCount: number;
  followsCount: number;
  postsCount: number;
  indexedAt: string;
  labels?: ATProtoLabel[];
}

export interface ATProtoLabel {
  src: string;
  uri: string;
  val: string;
  neg?: boolean;
  cts: string;
}

export interface ATProtoPost {
  uri: string;
  cid: string;
  author: ATProtoProfile;
  record: ATProtoPostRecord;
  replyCount: number;
  repostCount: number;
  likeCount: number;
  quoteCount: number;
  indexedAt: string;
  labels?: ATProtoLabel[];
  embed?: ATProtoEmbed;
}

export interface ATProtoPostRecord {
  $type: 'app.bsky.feed.post';
  text: string;
  createdAt: string;
  langs?: string[];
  facets?: ATProtoFacet[];
  reply?: ATProtoReplyRef;
  embed?: ATProtoEmbedRecord;
}

export interface ATProtoFacet {
  index: { byteStart: number; byteEnd: number };
  features: ATProtoFacetFeature[];
}

export type ATProtoFacetFeature =
  | { $type: 'app.bsky.richtext.facet#mention'; did: string }
  | { $type: 'app.bsky.richtext.facet#link'; uri: string }
  | { $type: 'app.bsky.richtext.facet#tag'; tag: string };

export interface ATProtoReplyRef {
  root: { uri: string; cid: string };
  parent: { uri: string; cid: string };
}

export interface ATProtoEmbedRecord {
  $type: string;
  images?: ATProtoImage[];
  external?: ATProtoExternal;
  record?: { uri: string; cid: string };
}

export interface ATProtoEmbed {
  $type: string;
  images?: ATProtoImageView[];
  external?: ATProtoExternalView;
  record?: ATProtoRecordView;
}

export interface ATProtoImage {
  alt: string;
  image: { $type: 'blob'; ref: { $link: string }; mimeType: string; size: number };
  aspectRatio?: { width: number; height: number };
}

export interface ATProtoImageView {
  thumb: string;
  fullsize: string;
  alt: string;
  aspectRatio?: { width: number; height: number };
}

export interface ATProtoExternal {
  uri: string;
  title: string;
  description: string;
  thumb?: { $type: 'blob'; ref: { $link: string }; mimeType: string; size: number };
}

export interface ATProtoExternalView {
  uri: string;
  title: string;
  description: string;
  thumb?: string;
}

export interface ATProtoRecordView {
  $type: string;
  uri: string;
  cid: string;
  author: ATProtoProfile;
  value: ATProtoPostRecord;
  indexedAt: string;
}

// DID Resolution

export interface DIDDocument {
  id: string;
  alsoKnownAs: string[];
  verificationMethod: DIDVerificationMethod[];
  service: DIDService[];
}

export interface DIDVerificationMethod {
  id: string;
  type: string;
  controller: string;
  publicKeyMultibase?: string;
}

export interface DIDService {
  id: string;
  type: string;
  serviceEndpoint: string;
}

export type DIDMethod = 'plc' | 'web';

// Bluesky Feed Types

export interface BlueskyFeed {
  uri: string;
  cid: string;
  did: string;
  creator: ATProtoProfile;
  displayName: string;
  description?: string;
  avatar?: string;
  likeCount: number;
  indexedAt: string;
}

export interface BlueskyTimeline {
  cursor?: string;
  feed: BlueskyFeedItem[];
}

export interface BlueskyFeedItem {
  post: ATProtoPost;
  reply?: {
    root: ATProtoPost;
    parent: ATProtoPost;
  };
  reason?: BlueskyFeedReason;
}

export type BlueskyFeedReason = {
  $type: 'app.bsky.feed.defs#reasonRepost';
  by: ATProtoProfile;
  indexedAt: string;
};

// Utility Functions

export function parseATUri(uri: string): { repo: string; collection: string; rkey: string } | null {
  const match = uri.match(/^at:\/\/([^/]+)\/([^/]+)\/([^/]+)$/);
  if (!match) return null;
  return { repo: match[1], collection: match[2], rkey: match[3] };
}

export function buildATUri(repo: string, collection: string, rkey: string): string {
  return `at://${repo}/${collection}/${rkey}`;
}

export function getPostUrl(handle: string, uri: string): string {
  const parsed = parseATUri(uri);
  const rkey = parsed?.rkey || uri.split('/').pop() || '';
  return `https://bsky.app/profile/${handle}/post/${rkey}`;
}

export function getProfileUrl(handle: string): string {
  return `https://bsky.app/profile/${handle}`;
}

export function getDIDMethod(did: string): DIDMethod | null {
  if (did.startsWith('did:plc:')) return 'plc';
  if (did.startsWith('did:web:')) return 'web';
  return null;
}

export function extractMentions(text: string): string[] {
  const mentionRegex = /@([a-zA-Z0-9]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?\.)+[a-zA-Z]([a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?/g;
  return (text.match(mentionRegex) || []).map((m) => m.slice(1));
}

export function extractHashtags(text: string): string[] {
  const hashtagRegex = /#[a-zA-Z][a-zA-Z0-9_]*/g;
  return (text.match(hashtagRegex) || []).map((h) => h.slice(1));
}

export function extractUrls(text: string): string[] {
  const urlRegex = /https?:\/\/[^\s<>[\]()]+/g;
  return text.match(urlRegex) || [];
}

// Convert AT Protocol post to Canonical format
export function atProtoToCanonical(post: ATProtoPost): CanonicalPost {
  return {
    id: post.uri,
    content: post.record.text,
    created_at: post.record.createdAt,
    author: {
      id: post.author.did,
      handle: post.author.handle,
      displayName: post.author.displayName || post.author.handle,
      avatar: post.author.avatar,
    },
    media: post.embed?.images?.map((img) => ({
      type: 'image' as const,
      url: img.fullsize,
      alt_text: img.alt,
      width: img.aspectRatio?.width,
      height: img.aspectRatio?.height,
    })),
    metrics: {
      likes: post.likeCount,
      reposts: post.repostCount,
      replies: post.replyCount,
      quotes: post.quoteCount,
    },
    reply_to: post.record.reply?.parent.uri,
    language: post.record.langs?.[0],
    labels: post.labels?.map((l) => l.val),
  };
}

// Convert Canonical post to AT Protocol record
export function canonicalToATProto(post: Partial<CanonicalPost>): Partial<ATProtoPostRecord> {
  const text = post.content || '';
  const facets: ATProtoFacet[] = [];

  // Extract mentions
  const mentions = extractMentions(text);
  mentions.forEach((handle) => {
    const start = text.indexOf(`@${handle}`);
    if (start !== -1) {
      facets.push({
        index: {
          byteStart: new TextEncoder().encode(text.slice(0, start)).length,
          byteEnd: new TextEncoder().encode(text.slice(0, start + handle.length + 1)).length,
        },
        features: [{ $type: 'app.bsky.richtext.facet#mention', did: '' }], // DID needs resolution
      });
    }
  });

  // Extract URLs
  const urls = extractUrls(text);
  urls.forEach((url) => {
    const start = text.indexOf(url);
    if (start !== -1) {
      facets.push({
        index: {
          byteStart: new TextEncoder().encode(text.slice(0, start)).length,
          byteEnd: new TextEncoder().encode(text.slice(0, start + url.length)).length,
        },
        features: [{ $type: 'app.bsky.richtext.facet#link', uri: url }],
      });
    }
  });

  // Extract hashtags
  const hashtags = extractHashtags(text);
  hashtags.forEach((tag) => {
    const start = text.indexOf(`#${tag}`);
    if (start !== -1) {
      facets.push({
        index: {
          byteStart: new TextEncoder().encode(text.slice(0, start)).length,
          byteEnd: new TextEncoder().encode(text.slice(0, start + tag.length + 1)).length,
        },
        features: [{ $type: 'app.bsky.richtext.facet#tag', tag }],
      });
    }
  });

  return {
    $type: 'app.bsky.feed.post',
    text,
    createdAt: post.created_at || new Date().toISOString(),
    langs: post.language ? [post.language] : undefined,
    facets: facets.length > 0 ? facets : undefined,
  };
}

// API Client wrapper for Bluesky endpoints
class BlueskyApiClient {
  private async request<T>(endpoint: string, options: RequestInit = {}): Promise<T> {
    const response = await fetch(`/api/v1/bluesky${endpoint}`, {
      ...options,
      headers: {
        'Content-Type': 'application/json',
        ...options.headers,
      },
      credentials: 'include',
    });

    const data = await response.json();

    if (!response.ok) {
      throw new Error(data.error || `HTTP ${response.status}`);
    }

    return data.data;
  }

  async getProfile(handle: string): Promise<ATProtoProfile> {
    return this.request(`/profile/${encodeURIComponent(handle)}`);
  }

  async getTimeline(cursor?: string, limit = 50): Promise<BlueskyTimeline> {
    const params = new URLSearchParams();
    if (cursor) params.set('cursor', cursor);
    params.set('limit', String(limit));
    return this.request(`/timeline?${params.toString()}`);
  }

  async getPost(uri: string): Promise<ATProtoPost> {
    return this.request(`/post/${encodeURIComponent(uri)}`);
  }

  async createPost(text: string, replyTo?: { uri: string; cid: string }): Promise<{ uri: string; cid: string }> {
    return this.request('/post', {
      method: 'POST',
      body: JSON.stringify({ text, reply_to: replyTo }),
    });
  }

  async deletePost(uri: string): Promise<{ deleted: boolean }> {
    return this.request(`/post/${encodeURIComponent(uri)}`, { method: 'DELETE' });
  }

  async like(uri: string, cid: string): Promise<{ uri: string }> {
    return this.request('/like', {
      method: 'POST',
      body: JSON.stringify({ uri, cid }),
    });
  }

  async unlike(likeUri: string): Promise<{ deleted: boolean }> {
    return this.request('/unlike', {
      method: 'POST',
      body: JSON.stringify({ like_uri: likeUri }),
    });
  }

  async repost(uri: string, cid: string): Promise<{ uri: string }> {
    return this.request('/repost', {
      method: 'POST',
      body: JSON.stringify({ uri, cid }),
    });
  }

  async unrepost(repostUri: string): Promise<{ deleted: boolean }> {
    return this.request('/unrepost', {
      method: 'POST',
      body: JSON.stringify({ repost_uri: repostUri }),
    });
  }

  async resolveHandle(identifier: string): Promise<ATProtoDID> {
    return this.request(`/resolve/${encodeURIComponent(identifier)}`);
  }

  async getPopularFeeds(limit = 25, cursor?: string): Promise<{ cursor?: string; feeds: BlueskyFeed[] }> {
    const params = new URLSearchParams();
    params.set('limit', String(limit));
    if (cursor) params.set('cursor', cursor);
    return this.request(`/feeds/popular?${params.toString()}`);
  }

  async getAuthorFeed(handle: string, cursor?: string, limit = 50): Promise<BlueskyTimeline> {
    const params = new URLSearchParams();
    if (cursor) params.set('cursor', cursor);
    params.set('limit', String(limit));
    return this.request(`/author/${encodeURIComponent(handle)}/feed?${params.toString()}`);
  }
}

export const blueskyApi = new BlueskyApiClient();

// Hooks

export function useBlueskyProfile(handle?: string) {
  return useQuery<ATProtoProfile>({
    queryKey: ['bluesky-profile', handle],
    queryFn: () => blueskyApi.getProfile(handle!),
    enabled: !!handle,
  });
}

export function useBlueskyTimeline(cursor?: string) {
  return useQuery<BlueskyTimeline>({
    queryKey: ['bluesky-timeline', cursor],
    queryFn: () => blueskyApi.getTimeline(cursor),
  });
}

export function useBlueskyPost(uri?: string) {
  return useQuery<ATProtoPost>({
    queryKey: ['bluesky-post', uri],
    queryFn: () => blueskyApi.getPost(uri!),
    enabled: !!uri,
  });
}

export function useResolveDID() {
  return useMutation({
    mutationFn: (identifier: string) => blueskyApi.resolveHandle(identifier),
  });
}

export function useCreateBlueskyPost() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (post: Partial<CanonicalPost>): Promise<PlatformMapping> => {
      const response = await blueskyApi.createPost(post.content || '');
      // Extract DID from URI (at://did:plc:xxx/app.bsky.feed.post/rkey)
      const parsed = parseATUri(response.uri);
      const did = parsed?.repo || 'unknown';
      return {
        platform: 'bluesky',
        native_id: response.uri,
        url: getPostUrl(did, response.uri),
        status: 'synced',
        synced_at: new Date().toISOString(),
      };
    },
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['bluesky-timeline'] });
    },
  });
}

export function useDeleteBlueskyPost() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: (uri: string) => blueskyApi.deletePost(uri),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['bluesky-timeline'] });
    },
  });
}

export function useLikeBlueskyPost() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: ({ uri, cid }: { uri: string; cid: string }) => blueskyApi.like(uri, cid),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['bluesky-timeline'] });
    },
  });
}

export function useUnlikeBlueskyPost() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: (likeUri: string) => blueskyApi.unlike(likeUri),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['bluesky-timeline'] });
    },
  });
}

export function useRepostBluesky() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: ({ uri, cid }: { uri: string; cid: string }) => blueskyApi.repost(uri, cid),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['bluesky-timeline'] });
    },
  });
}

export function useUnrepostBluesky() {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: (repostUri: string) => blueskyApi.unrepost(repostUri),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['bluesky-timeline'] });
    },
  });
}

// Custom Feed Hooks

export function useBlueskyFeeds() {
  return useQuery<BlueskyFeed[]>({
    queryKey: ['bluesky-feeds'],
    queryFn: async () => {
      const response = await blueskyApi.getPopularFeeds();
      return response.feeds;
    },
  });
}

export function useBlueskyAuthorFeed(handle?: string, cursor?: string) {
  return useQuery<BlueskyTimeline>({
    queryKey: ['bluesky-author-feed', handle, cursor],
    queryFn: () => blueskyApi.getAuthorFeed(handle!, cursor),
    enabled: !!handle,
  });
}

File: app/web/api/v1/bluesky.py
"""
Bluesky API endpoints for direct platform interactions.

Provides endpoints for:
- Profile fetching
- Timeline/feed retrieval
- Post creation, deletion
- Likes and reposts
- DID resolution
"""

from flask import Blueprint, request
from atproto import Client
from atproto.exceptions import AtProtocolError

from app.auth.auth_decorators import require_auth
from app.auth.credential_manager import CredentialManager
from app.web.api.v1.responses import api_response, api_error
from app.core.logger import setup_logger

logger = setup_logger(__name__)

bluesky_bp = Blueprint("bluesky", __name__, url_prefix="/bluesky")


def _get_bluesky_client(user_id: int) -> Client:
    """Get authenticated Bluesky client for user."""
    cm = CredentialManager()
    creds = cm.get_credentials(user_id, "bluesky", "api")

    if not creds:
        raise ValueError("No Bluesky credentials found")

    client = Client()
    client.login(creds["identifier"], creds["password"])
    return client


def _format_profile(profile) -> dict:
    """Format AT Protocol profile to API response."""
    return {
        "did": profile.did,
        "handle": profile.handle,
        "displayName": getattr(profile, "display_name", None) or profile.handle,
        "description": getattr(profile, "description", None),
        "avatar": getattr(profile, "avatar", None),
        "banner": getattr(profile, "banner", None),
        "followersCount": getattr(profile, "followers_count", 0),
        "followsCount": getattr(profile, "follows_count", 0),
        "postsCount": getattr(profile, "posts_count", 0),
        "indexedAt": getattr(profile, "indexed_at", None),
        "labels": [{"val": l.val, "src": l.src} for l in (getattr(profile, "labels", None) or [])],
    }


def _format_post(post) -> dict:
    """Format AT Protocol post to API response."""
    record = post.record if hasattr(post, "record") else {}
    author = post.author if hasattr(post, "author") else None
    embed = getattr(post, "embed", None)

    formatted = {
        "uri": post.uri,
        "cid": post.cid,
        "author": _format_profile(author) if author else None,
        "record": {
            "$type": "app.bsky.feed.post",
            "text": getattr(record, "text", "") if hasattr(record, "text") else record.get("text", ""),
            "createdAt": getattr(record, "created_at", None) if hasattr(record, "created_at") else record.get("createdAt"),
            "langs": getattr(record, "langs", None) if hasattr(record, "langs") else record.get("langs"),
        },
        "replyCount": getattr(post, "reply_count", 0),
        "repostCount": getattr(post, "repost_count", 0),
        "likeCount": getattr(post, "like_count", 0),
        "quoteCount": getattr(post, "quote_count", 0),
        "indexedAt": getattr(post, "indexed_at", None),
        "labels": [{"val": l.val, "src": l.src} for l in (getattr(post, "labels", None) or [])],
    }

    # Handle embeds
    if embed:
        formatted["embed"] = _format_embed(embed)

    return formatted


def _format_embed(embed) -> dict:
    """Format post embed (images, external links, etc.)."""
    result = {"$type": getattr(embed, "py_type", "unknown")}

    # Images
    if hasattr(embed, "images") and embed.images:
        result["images"] = [
            {
                "thumb": img.thumb,
                "fullsize": img.fullsize,
                "alt": getattr(img, "alt", ""),
                "aspectRatio": {
                    "width": img.aspect_ratio.width,
                    "height": img.aspect_ratio.height,
                } if hasattr(img, "aspect_ratio") and img.aspect_ratio else None,
            }
            for img in embed.images
        ]

    # External links
    if hasattr(embed, "external") and embed.external:
        ext = embed.external
        result["external"] = {
            "uri": ext.uri,
            "title": getattr(ext, "title", ""),
            "description": getattr(ext, "description", ""),
            "thumb": getattr(ext, "thumb", None),
        }

    return result


@bluesky_bp.route("/profile/<handle>", methods=["GET"])
@require_auth
def get_profile(handle: str):
    """Get Bluesky profile by handle or DID."""
    try:
        client = _get_bluesky_client(request.user_id)
        profile = client.get_profile(handle)
        return api_response(_format_profile(profile))
    except ValueError as e:
        return api_error(str(e), 400)
    except AtProtocolError as e:
        logger.error(f"Bluesky API error fetching profile: {e}")
        return api_error(f"Bluesky error: {str(e)}", 502)
    except Exception as e:
        logger.error(f"Error fetching Bluesky profile: {e}")
        return api_error("Failed to fetch profile", 500)


@bluesky_bp.route("/timeline", methods=["GET"])
@require_auth
def get_timeline():
    """Get authenticated user's home timeline."""
    cursor = request.args.get("cursor")
    limit = min(int(request.args.get("limit", 50)), 100)

    try:
        client = _get_bluesky_client(request.user_id)
        response = client.get_timeline(cursor=cursor, limit=limit)

        feed = []
        for item in response.feed:
            feed_item = {
                "post": _format_post(item.post),
            }
            if hasattr(item, "reply") and item.reply:
                feed_item["reply"] = {
                    "root": _format_post(item.reply.root) if item.reply.root else None,
                    "parent": _format_post(item.reply.parent) if item.reply.parent else None,
                }
            if hasattr(item, "reason") and item.reason:
                feed_item["reason"] = {
                    "$type": getattr(item.reason, "py_type", "unknown"),
                    "by": _format_profile(item.reason.by) if hasattr(item.reason, "by") else None,
                    "indexedAt": getattr(item.reason, "indexed_at", None),
                }
            feed.append(feed_item)

        return api_response({
            "cursor": response.cursor,
            "feed": feed,
        })
    except ValueError as e:
        return api_error(str(e), 400)
    except AtProtocolError as e:
        logger.error(f"Bluesky API error fetching timeline: {e}")
        return api_error(f"Bluesky error: {str(e)}", 502)
    except Exception as e:
        logger.error(f"Error fetching Bluesky timeline: {e}")
        return api_error("Failed to fetch timeline", 500)


@bluesky_bp.route("/feed/<path:feed_uri>", methods=["GET"])
@require_auth
def get_feed(feed_uri: str):
    """Get posts from a custom feed."""
    cursor = request.args.get("cursor")
    limit = min(int(request.args.get("limit", 50)), 100)

    try:
        client = _get_bluesky_client(request.user_id)
        response = client.app.bsky.feed.get_feed(
            {"feed": feed_uri, "cursor": cursor, "limit": limit}
        )

        feed = [{"post": _format_post(item.post)} for item in response.feed]

        return api_response({
            "cursor": response.cursor,
            "feed": feed,
        })
    except ValueError as e:
        return api_error(str(e), 400)
    except AtProtocolError as e:
        logger.error(f"Bluesky API error fetching feed: {e}")
        return api_error(f"Bluesky error: {str(e)}", 502)
    except Exception as e:
        logger.error(f"Error fetching Bluesky feed: {e}")
        return api_error("Failed to fetch feed", 500)


@bluesky_bp.route("/post/<path:uri>", methods=["GET"])
@require_auth
def get_post(uri: str):
    """Get a single post by URI."""
    try:
        client = _get_bluesky_client(request.user_id)
        # Parse URI to get repo and rkey
        parts = uri.replace("at://", "").split("/")
        if len(parts) < 3:
            return api_error("Invalid post URI", 400)

        repo, collection, rkey = parts[0], parts[1], parts[2]
        response = client.get_post(rkey, repo)

        return api_response(_format_post(response))
    except ValueError as e:
        return api_error(str(e), 400)
    except AtProtocolError as e:
        logger.error(f"Bluesky API error fetching post: {e}")
        return api_error(f"Bluesky error: {str(e)}", 502)
    except Exception as e:
        logger.error(f"Error fetching Bluesky post: {e}")
        return api_error("Failed to fetch post", 500)


@bluesky_bp.route("/post", methods=["POST"])
@require_auth
def create_post():
    """Create a new post on Bluesky."""
    data = request.get_json() or {}
    text = data.get("text", "").strip()

    if not text:
        return api_error("Post text is required", 400)

    if len(text) > 300:
        return api_error("Post text exceeds 300 character limit", 400)

    reply_to = data.get("reply_to")  # Optional: {uri, cid}

    try:
        client = _get_bluesky_client(request.user_id)

        # Build reply reference if provided
        reply_ref = None
        if reply_to and reply_to.get("uri") and reply_to.get("cid"):
            from atproto import models
            parent = models.create_strong_ref(
                models.ComAtprotoRepoStrongRef.Main(
                    uri=reply_to["uri"],
                    cid=reply_to["cid"],
                )
            )
            # For reply, root is the same as parent unless specified
            root_data = reply_to.get("root", reply_to)
            root = models.create_strong_ref(
                models.ComAtprotoRepoStrongRef.Main(
                    uri=root_data["uri"],
                    cid=root_data["cid"],
                )
            )
            reply_ref = models.AppBskyFeedPost.ReplyRef(parent=parent, root=root)

        response = client.send_post(text=text, reply_to=reply_ref)

        return api_response({
            "uri": response.uri,
            "cid": response.cid,
        }, 201)
    except ValueError as e:
        return api_error(str(e), 400)
    except AtProtocolError as e:
        logger.error(f"Bluesky API error creating post: {e}")
        return api_error(f"Bluesky error: {str(e)}", 502)
    except Exception as e:
        logger.error(f"Error creating Bluesky post: {e}")
        return api_error("Failed to create post", 500)


@bluesky_bp.route("/post/<path:uri>", methods=["DELETE"])
@require_auth
def delete_post(uri: str):
    """Delete a post by URI."""
    try:
        client = _get_bluesky_client(request.user_id)
        client.delete_post(uri)
        return api_response({"deleted": True})
    except ValueError as e:
        return api_error(str(e), 400)
    except AtProtocolError as e:
        logger.error(f"Bluesky API error deleting post: {e}")
        return api_error(f"Bluesky error: {str(e)}", 502)
    except Exception as e:
        logger.error(f"Error deleting Bluesky post: {e}")
        return api_error("Failed to delete post", 500)


@bluesky_bp.route("/like", methods=["POST"])
@require_auth
def like_post():
    """Like a post."""
    data = request.get_json() or {}
    uri = data.get("uri")
    cid = data.get("cid")

    if not uri or not cid:
        return api_error("uri and cid are required", 400)

    try:
        client = _get_bluesky_client(request.user_id)
        response = client.like(uri, cid)
        return api_response({"uri": response.uri})
    except ValueError as e:
        return api_error(str(e), 400)
    except AtProtocolError as e:
        logger.error(f"Bluesky API error liking post: {e}")
        return api_error(f"Bluesky error: {str(e)}", 502)
    except Exception as e:
        logger.error(f"Error liking Bluesky post: {e}")
        return api_error("Failed to like post", 500)


@bluesky_bp.route("/unlike", methods=["POST"])
@require_auth
def unlike_post():
    """Remove like from a post."""
    data = request.get_json() or {}
    like_uri = data.get("like_uri")

    if not like_uri:
        return api_error("like_uri is required", 400)

    try:
        client = _get_bluesky_client(request.user_id)
        client.delete_like(like_uri)
        return api_response({"deleted": True})
    except ValueError as e:
        return api_error(str(e), 400)
    except AtProtocolError as e:
        logger.error(f"Bluesky API error unliking post: {e}")
        return api_error(f"Bluesky error: {str(e)}", 502)
    except Exception as e:
        logger.error(f"Error unliking Bluesky post: {e}")
        return api_error("Failed to unlike post", 500)


@bluesky_bp.route("/repost", methods=["POST"])
@require_auth
def repost():
    """Repost a post."""
    data = request.get_json() or {}
    uri = data.get("uri")
    cid = data.get("cid")

    if not uri or not cid:
        return api_error("uri and cid are required", 400)

    try:
        client = _get_bluesky_client(request.user_id)
        response = client.repost(uri, cid)
        return api_response({"uri": response.uri})
    except ValueError as e:
        return api_error(str(e), 400)
    except AtProtocolError as e:
        logger.error(f"Bluesky API error reposting: {e}")
        return api_error(f"Bluesky error: {str(e)}", 502)
    except Exception as e:
        logger.error(f"Error reposting on Bluesky: {e}")
        return api_error("Failed to repost", 500)


@bluesky_bp.route("/unrepost", methods=["POST"])
@require_auth
def unrepost():
    """Remove repost."""
    data = request.get_json() or {}
    repost_uri = data.get("repost_uri")

    if not repost_uri:
        return api_error("repost_uri is required", 400)

    try:
        client = _get_bluesky_client(request.user_id)
        client.delete_repost(repost_uri)
        return api_response({"deleted": True})
    except ValueError as e:
        return api_error(str(e), 400)
    except AtProtocolError as e:
        logger.error(f"Bluesky API error removing repost: {e}")
        return api_error(f"Bluesky error: {str(e)}", 502)
    except Exception as e:
        logger.error(f"Error removing repost on Bluesky: {e}")
        return api_error("Failed to remove repost", 500)


@bluesky_bp.route("/resolve/<identifier>", methods=["GET"])
@require_auth
def resolve_handle(identifier: str):
    """Resolve a handle to DID or vice versa."""
    try:
        client = _get_bluesky_client(request.user_id)

        if identifier.startswith("did:"):
            # Resolve DID to handle
            profile = client.get_profile(identifier)
            return api_response({
                "did": profile.did,
                "handle": profile.handle,
                "resolved_at": None,  # API doesn't provide this
            })
        else:
            # Resolve handle to DID
            response = client.resolve_handle(identifier)
            return api_response({
                "did": response.did,
                "handle": identifier,
                "resolved_at": None,
            })
    except ValueError as e:
        return api_error(str(e), 400)
    except AtProtocolError as e:
        logger.error(f"Bluesky API error resolving identifier: {e}")
        return api_error(f"Bluesky error: {str(e)}", 502)
    except Exception as e:
        logger.error(f"Error resolving Bluesky identifier: {e}")
        return api_error("Failed to resolve identifier", 500)


@bluesky_bp.route("/feeds/popular", methods=["GET"])
@require_auth
def get_popular_feeds():
    """Get popular/suggested feeds."""
    limit = min(int(request.args.get("limit", 25)), 100)
    cursor = request.args.get("cursor")

    try:
        client = _get_bluesky_client(request.user_id)
        response = client.app.bsky.feed.get_suggested_feeds(
            {"limit": limit, "cursor": cursor}
        )

        feeds = []
        for feed in response.feeds:
            feeds.append({
                "uri": feed.uri,
                "cid": feed.cid,
                "did": feed.did,
                "creator": _format_profile(feed.creator) if feed.creator else None,
                "displayName": feed.display_name,
                "description": getattr(feed, "description", None),
                "avatar": getattr(feed, "avatar", None),
                "likeCount": getattr(feed, "like_count", 0),
                "indexedAt": getattr(feed, "indexed_at", None),
            })

        return api_response({
            "cursor": response.cursor,
            "feeds": feeds,
        })
    except ValueError as e:
        return api_error(str(e), 400)
    except AtProtocolError as e:
        logger.error(f"Bluesky API error fetching feeds: {e}")
        return api_error(f"Bluesky error: {str(e)}", 502)
    except Exception as e:
        logger.error(f"Error fetching Bluesky feeds: {e}")
        return api_error("Failed to fetch feeds", 500)


@bluesky_bp.route("/author/<handle>/feed", methods=["GET"])
@require_auth
def get_author_feed(handle: str):
    """Get posts from a specific author."""
    cursor = request.args.get("cursor")
    limit = min(int(request.args.get("limit", 50)), 100)

    try:
        client = _get_bluesky_client(request.user_id)
        response = client.get_author_feed(handle, cursor=cursor, limit=limit)

        feed = [{"post": _format_post(item.post)} for item in response.feed]

        return api_response({
            "cursor": response.cursor,
            "feed": feed,
        })
    except ValueError as e:
        return api_error(str(e), 400)
    except AtProtocolError as e:
        logger.error(f"Bluesky API error fetching author feed: {e}")
        return api_error(f"Bluesky error: {str(e)}", 502)
    except Exception as e:
        logger.error(f"Error fetching Bluesky author feed: {e}")
        return api_error("Failed to fetch author feed", 500)
